riscv/plic.cc

```diff
@@ -66,21 +67,28 @@
 
 #define REG_SIZE                0x1000000
 
-plic_t::plic_t(std::vector<processor_t*>& procs, bool smode, uint32_t ndev)
-  : contexts(procs.size() * (smode ? 2 : 1)),
-  num_ids(ndev + 1), num_ids_word(((ndev + 1) + (32 - 1)) / 32),
+plic_t::plic_t(sim_t* sim, uint32_t ndev)
+  : num_ids(ndev + 1), num_ids_word(((ndev + 1) + (32 - 1)) / 32),
   max_prio((1UL << PLIC_PRIO_BITS) - 1) 
 {
-  size_t contexts_per_hart = smode ? 2 : 1;
-
   memset(priority, 0, sizeof(priority));
   memset(level, 0, sizeof(level));
 
-  for (size_t i = 0; i < contexts.size(); i++) {
-    plic_context_t* c = &contexts[i];
-    memset(c, 0, sizeof(*c));
-    c->proc = procs[i / contexts_per_hart];
-    c->mmode = (i % contexts_per_hart == 0);
+  // PLIC contexts are contiguous in memory even if harts are discontiguous.
+  for (const auto& [hart_id, hart] : sim->get_harts()) {
+    contexts.push_back(plic_context_t());
+    plic_context_t* m_context = &contexts.back();
+    memset(m_context, 0, sizeof(*m_context));
```

Comment:
Yuck. Why not let `plic_context_t`'s constructor initialize all the fields? Including `proc` and `mmode`

Reply:
Yes, that’s better still.